## 异常处理

### 1、抛出异常：throw关键字

在方法中使用throw关键字抛出指定的异常

- 在创建函数时，一定要对传递的参数进行**合法性判断**

```Java
throw new NullPointerException("传入参数为空");	// 抛出空指针异常，并输出说明
```

#### 注意点

- throw必须写在方法的内部

- throw后创建的必须是Exception或Exception的子类对象
- throw抛出指定的异常对象，则必须进行处理
  - throw后创建的是RunTimeException或RunTimeException的子类对象，则可以不处理，默认交给JVM处理（即打印异常+中断程序）
  - throw后创建的是编译异常，则必须处理（要么throws、要么try..catch）

### 常见异常

1、``NullPointerException`空指针异常

2、``AaaryIndexOutOfBoundEception`数组越界异常

#### Objects非空判断

对空指针异常判断时，可使用Objects类的方法简化代码

```Java
Objects.requireNotNull(E e);

/*
if(e == null){
	throw new NullPointerException("传入参数为空");
}
*/
```

### 2、异常处理

#### throws关键字：声明异常而不处理

在抛出异常的方法名最后生命方法中抛出的所有异常，将异常交给调用该方法的调用者处理

- 抛出多少异常，就要声明多少异常
- 抛出的异常中父类、子类都有，只声明父类异常
- 异常声明需要`往上传递`，只要方法中会抛出异常，都要进行声明

#### try...catch：抛出异常也进行处理

执行try中的程序，若catch捕捉到异常则抛出，但继续执行程序

```Java
try{
	...
}catch(NullPointerException e){
	System.out.println("空指针异常");
}
```

#### try...catch...finally

finally一般用于资源释放（资源回收），无论try中的代码是否有异常，最终都会执行

#### Throwable类打印异常信息的三个方法

- getMessage()：打印异常的简短说明
- toString()：打印异常的详细信息，与直接输出异常时一样的
- printStackTrace：打印的异常信息最详细，JVM默认调用此方法

### 2、异常处理注意事项

finally若有return语句，则函数使用时，finally中的return始终会覆盖try中的return，要尽量避免这种情况

#### 子父类异常的声明

父类方法声明异常，子类对该方法重写时，有以下几种情况

- 抛出和父类相同的异常
- 抛出父类异常的子类
- 不抛出异常

父类方法没有声明异常，子类重写该方法时，也不能声明，只能对异常做处理，即try...catch...finally

#### 异常处理的3 种方式

- 多个异常分别处理，即一次抛出对应一次处理

```
try{
	...
}catch(){
	...
}

try{
	...
}catch(){
	...
}
```

- 多个异常一次抛出，多次处理

抛出具体的异常，分别处理

当同时抛出子类和父类异常时，子类异常要写在父类异常之前，否则子类异常对象将不会被使用

```Java
try{
	...
}catch(子类异常 e){
	...
}catch(父类异常 e){
	...
}finally{
	...
}
```

- 多个异常一次抛出，一次处理

抛出Exception异常，即接收任何子类异常，一次处理

```Java
try{
	...
}ctach(Exception e){
	...
}
```

